state-chain/pallets/cf-asset-balances/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![doc = include_str!("../../cf-doc-head.md")]
+
+use cf_chains::{AnyChain, ForeignChain, ForeignChainAddress};
+use cf_primitives::{Asset, AssetAmount};
+use cf_runtime_utilities::log_or_panic;
+use cf_traits::{
+	impl_pallet_safe_mode, AssetWithholding, Chainflip, EgressApi, KeyProvider, LiabilityTracker,
+	ScheduledEgressDetails,
+};
+use frame_support::{
+	pallet_prelude::*, sp_runtime::traits::Saturating, traits::DefensiveSaturating,
+};
+use serde::{Deserialize, Serialize};
+use sp_std::{collections::btree_map::BTreeMap, vec, vec::Vec};
+
+pub use pallet::*;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub const PALLET_VERSION: StorageVersion = StorageVersion::new(0);
+
+// TODO: Figure out adequate values for these.
+pub const MAX_ETH_REFUND_PER_EPOCH: usize = 25;
```

Comment:
@kylezs and @albert-llimos happy to get some opinion on what would be a reasonable value for this.

Reply:
Why do we need a max? If we have the ability to refund them and we've kept the funds aside, why not do it?

Reply:
Primarily because we assume that for the first refund, the batch could become quite massive ~250. I think we fear a bit the unknowns of this.

Reply:
Ooh, it's not the MAX_ETH_REFUND, but it's `MAX_REFUNDED_VALIDATORS_ETH_PER_EPOCH` or something like that. 

Even still, if we're only refunding 25, then it'll take over 2 weeks to refund everyone, seems too long. Keen to hear @albert-llimos 's opinion here on the batch size. Ideally we can just submit everything in one batch but maybe there is a practical limitation there.

Reply:
Dan looped me into the conversation here: https://github.com/chainflip-io/chainflip-backend/pull/4962#pullrequestreview-2160793703.
The batch size for ~250 transfers might be problematic for gas limits. According to some tests we shouldn't be close to the gas limit but it's still "risky" to do that massive tx.
It looked like the best solution was to either limit it (like done here) or to use the new logic added on Solana to limit the amount of transfers that can be put together in a single `AllBatch`. My preference was to cherry pick that logic for Solana (small) into here and set it to a reasonable amount of transfers so it would automatically be split up. Especially since we want to include a limit for Ethereum anyway. That's what I last told to Dan  but I'm not sure if there was any other new information after that.
That would still be me preferred solution.

Reply:
FWIW, we should be able to easily get away with at least 50 or even 100 refunds in one go, so if we want to just increase that hardcoded number instead of adding new logic for simplicity, that's also fine. Then the refund will take a shorter time.

Reply:
Ok makes sense. In that case we definitely want to split it up. Makes sense. 
Not sure about splitting it per epoch though, only 25 per epoch would take forever. We could just construct all the batches up front and schedule them 10 SC blocks apart or something, rather than wait potentially over 2 weeks before the next validators can get their refund. 

Where can I see the Solana AllBatch limit code? If we can code share that would be nice.

Reply:
Ok that looks fairly involved. Given the size of this PR already, I'd rather make the batch 50 and create an issue to address it in a better way.

Reply:
I think the design would work quite well and easy for the future use-case where we don't collect fees other more than half a year. I would not make the implementation unnecessary complicated for a single edge case. I think an epoch is 3 days, at the moment, right? If you say 100 should be doable, and we go with let's say 75 per epoch (to be safe) all current validators should be refunded in about 10 days, right? Would sound acceptable to me.

Reply:
https://linear.app/chainflip/issue/PRO-1513/better-refund-batching

Reply:
If I think more about it, I think it would make sense to split it up by blocks over the entire epoch. I don't think we should hit 75-100 on a single epoch right now, but it's actually not so unrealistic (should check the numbers) especially when we expect volume to increase in the future.

Reply:
Yeah there are plenty of ways we could do it, but I think we should think about it separately. Doing 50 or so per epoch for now is good enough, and we can continue thinking about it after this is in.